nimare/meta/cbmr.py

```diff
@@ -0,0 +1,903 @@
+            Dataset to analyze.
+        """
+        init_weight_kwargs = {
+            'groups': self.groups,
+            'spatial_coef_dim': self.inputs_["coef_spline_bases"].shape[1],
+            'moderators_coef_dim': len(self.moderators) if self.moderators else None,
+        }
+        self.model.init_weights(**init_weight_kwargs)
+
+        moderators_by_group = self.inputs_["moderators_by_group"] if self.moderators else None
+        self.model.fit(self.inputs_["coef_spline_bases"], moderators_by_group, self.inputs_["foci_per_voxel"], self.inputs_["foci_per_study"])
+
+        maps, tables = self.model.summary()
+
+        return maps, tables
+
+
+class CBMRInference(object):
+    """Statistical inference on outcomes (intensity estimation and study-level
+    moderator regressors) of CBMR.
+
+    Parameters
+    ----------
+    CBMRResults : :obj:`~nimare.results.MetaResult`
+        Results of optimized regression coefficients of CBMR, as well as their
+        standard error in `tables`. Results of estimated spatial intensity function
+        (per study) in `maps`.
+    t_con_group : :obj:`~bool` or obj:`~list` or obj:`~None`, optional
+        Contrast matrix for homogeneity test or group comparison on estimated spatial
+        intensity function.
+        For boolean inputs, no statistical inference will be conducted for spatial intensity
+        if `t_con_group` is False, and spatial homogeneity test for groupwise intensity
+        function will be conducted if `t_con_group` is True.
+        For list inputs, generialized linear hypothesis (GLH) testing will be conducted for
+        each element independently. We also allow any element of `t_con_group` in list type,
+        which represents GLH is conducted for all contrasts in this element simultaneously.
+        Default is homogeneity test on group-wise estimated intensity function.
+    t_con_moderators : :obj:`~bool` or obj:`~list` or obj:`~None`, optional
+        Contrast matrix for testing the existence of one or more study-level moderator effects.
+        For boolean inputs, no statistical inference will be conducted for study-level moderators
+        if `t_con_moderators` is False, and statistical inference on the effect of each study-level
+        moderators will be conducted if `t_con_group` is True.
+        For list inputs, generialized linear hypothesis (GLH) testing will be conducted for
+        each element independently. We also allow any element of `t_con_moderators` in list type,
+        which represents GLH is conducted for all contrasts in this element simultaneously.
+        Default is statistical inference on the effect of each study-level moderators
+    device: :obj:`string`, optional
+        Device type ('cpu' or 'cuda') represents the device on which operations will be allocated.
+        Default is 'cpu'.
+    """
+    
+    def __init__(self, CBMRResults, t_con_group=None, t_con_moderator=None, device="cpu"): 
```

Comment:
Notes:

-  remove t_con_group and t_con_moderator from initialization
- find group and moderator names
  - store the names in two separate dictionaries (one for groups/one for moderators) where the keys are the names and the values are the index in the contrast array
- print out the dictionaries in a visually appealing way when initializing the class
- (need to flesh out more): create a function using the dictionaries to construct a contrast using the names of the groups/moderators
- create a function to run the contrasts

Reply:
expected API interaction:

```python
inference = CBMRInference(results, device='cuda')
```
output:
```
Group Reference in contrast array
"Schizophrenia_Yes = index 0"
"Schizophrenia_No = index 1"
...

Moderator Reference in contrast array
"sample_size = index 0"
...

```

```python
# convienence method to make basic contrasts using group/moderator names
contrast = inference.create_contrast(????, type="group") # or type="moderator"
contrast_result = inference.comput_contrast(contrast, type="group")
constrast_result.summary()
# display the statistics
```